File: src/tessera/url_rules.py
"""Static URL pattern rules: a fast deterministic gate before any LLM judge.

A two-tier outbound flow looks like this:

    1. URLRules.evaluate(url, method)
        -> ALLOW (allowlist hit)   -> short-circuit, skip everything below
        -> DENY  (denylist hit)    -> short-circuit, return rule_id
        -> NO_MATCH                -> fall through to the rest of the pipeline

    2. SSRFGuard.check_url(url)         (defense for "ALLOW" wasn't enough)
    3. LLMGuardrail / scanners          (semantic analysis on the body)

The point of this layer is to skip the expensive scanners and LLM judge
on the high-traffic predictable URLs: known-good API endpoints get an
immediate ALLOW, known-bad domains get an immediate DENY with a stable
rule_id in the audit log. Everything ambiguous falls through.

Pattern shapes
--------------
Three pattern types, evaluated in order: ``exact`` > ``prefix`` > ``glob``.
This ordering is meaningful: it gives the operator a way to carve out
exceptions from broad rules. Within each tier deny-wins-over-allow, so
adding a deny rule for ``/admin`` cannot be silently overridden by a
permissive ``/`` allow.

* ``exact``: string equality on the full URL.
* ``prefix``: the rule's pattern is a prefix of the URL.
* ``glob``: ``fnmatch``-style ``*`` and ``?`` wildcards on the full URL.

Method filter
-------------
Each rule may carry a method allowlist (``methods=["GET", "POST"]``).
If set, the rule only applies when the request's HTTP method is in the
list. ``methods=None`` (default) matches any method.

Why this is not a regex engine
------------------------------
Regex authoring is a footgun for security policy: catastrophic
backtracking, accidental anchoring, escaping mistakes. The three
shapes above cover everything operators reliably want to express
without giving them a way to ship a denial-of-service rule.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from fnmatch import fnmatchcase
from typing import Iterable


class RuleAction(StrEnum):
    ALLOW = "allow"
    DENY = "deny"


class RuleVerdict(StrEnum):
    """The gate's outcome on a single URL+method check."""

    ALLOW = "allow"
    DENY = "deny"
    NO_MATCH = "no_match"


class PatternKind(StrEnum):
    EXACT = "exact"
    PREFIX = "prefix"
    GLOB = "glob"


@dataclass(frozen=True)
class URLRule:
    """One static rule.

    Attributes:
        rule_id: Stable identifier for audit logs (``url.allow.github.api``,
            ``url.deny.internal.admin``, etc.). Operators choose the
            namespace.
        pattern: The string the rule matches against.
        kind: How to interpret ``pattern`` (exact / prefix / glob).
        action: ALLOW or DENY when this rule fires.
        methods: Optional uppercase HTTP method allowlist; None = any.
        description: Human-readable note for the audit log.
    """

    rule_id: str
    pattern: str
    kind: PatternKind = PatternKind.EXACT
    action: RuleAction = RuleAction.ALLOW
    methods: tuple[str, ...] | None = None
    description: str = ""

    def matches_method(self, method: str) -> bool:
        if self.methods is None:
            return True
        return method.upper() in self.methods

    def matches_url(self, url: str) -> bool:
        if self.kind == PatternKind.EXACT:
            return url == self.pattern
        if self.kind == PatternKind.PREFIX:
            return url.startswith(self.pattern)
        if self.kind == PatternKind.GLOB:
            return fnmatchcase(url, self.pattern)
        return False


@dataclass(frozen=True)
class URLDecision:
    """The gate's full response: verdict plus the rule that produced it."""

    verdict: RuleVerdict
    rule_id: str = ""
    description: str = ""
    method: str = ""
    url: str = ""

    @property
    def allowed(self) -> bool:
        return self.verdict == RuleVerdict.ALLOW


class URLRulesEngine:
    """Evaluate URLs against a list of static rules.

    Args:
        rules: Iterable of :class:`URLRule`. Order within a kind tier
            does not change the outcome (deny always wins within a tier);
            exact > prefix > glob across tiers.

    Behavior:
        * Empty rule set returns NO_MATCH for every URL.
        * Within a tier, if any DENY rule matches, the verdict is DENY.
        * Within a tier, if no DENY but at least one ALLOW matches, the
          verdict is ALLOW.
        * If no rule in the tier matches, fall through to the next tier.
        * If no tier matches, return NO_MATCH so the caller can run the
          slower checks.
    """

    name = "tessera.url_rules"

    def __init__(self, rules: Iterable[URLRule] = ()) -> None:
        self._exact: list[URLRule] = []
        self._prefix: list[URLRule] = []
        self._glob: list[URLRule] = []
        for rule in rules:
            self.add(rule)

    def add(self, rule: URLRule) -> None:
        if rule.kind == PatternKind.EXACT:
            self._exact.append(rule)
        elif rule.kind == PatternKind.PREFIX:
            self._prefix.append(rule)
        elif rule.kind == PatternKind.GLOB:
            self._glob.append(rule)

    def evaluate(self, url: str, method: str = "GET") -> URLDecision:
        """Walk the tiers and return the first decisive verdict."""
        for tier in (self._exact, self._prefix, self._glob):
            verdict = self._evaluate_tier(tier, url, method)
            if verdict is not None:
                return verdict
        return URLDecision(verdict=RuleVerdict.NO_MATCH, url=url, method=method)

    def _evaluate_tier(
        self,
        tier: list[URLRule],
        url: str,
        method: str,
    ) -> URLDecision | None:
        """Apply deny-wins-over-allow within a tier.

        Returns:
            A URLDecision when at least one rule in the tier matched
            both the URL and the method. None when no rule in the tier
            matched, signaling fall-through.
        """
        deny_hit: URLRule | None = None
        allow_hit: URLRule | None = None
        for rule in tier:
            if not rule.matches_url(url):
                continue
            if not rule.matches_method(method):
                continue
            if rule.action == RuleAction.DENY:
                deny_hit = rule
                break  # Deny wins; no need to look further in this tier.
            if rule.action == RuleAction.ALLOW and allow_hit is None:
                allow_hit = rule
        chosen = deny_hit or allow_hit
        if chosen is None:
            return None
        verdict = (
            RuleVerdict.DENY if chosen.action == RuleAction.DENY
            else RuleVerdict.ALLOW
        )
        return URLDecision(
            verdict=verdict,
            rule_id=chosen.rule_id,
            description=chosen.description,
            method=method.upper(),
            url=url,
        )

    @property
    def rule_count(self) -> int:
        return len(self._exact) + len(self._prefix) + len(self._glob)


__all__ = [
    "PatternKind",
    "RuleAction",
    "RuleVerdict",
    "URLDecision",
    "URLRule",
    "URLRulesEngine",
]

File: rust/crates/tessera-policy/src/url_rules.rs
//! Static URL pattern rules: a fast deterministic gate before any LLM
//! judge or DNS-resolving SSRF check.
//!
//! Mirrors `tessera.url_rules` from the Python reference. Three pattern
//! shapes evaluated in tier order (`exact > prefix > glob`); within each
//! tier, deny-wins-over-allow. A NO_MATCH falls through to the slower
//! checks (SSRF, scanners). Optional method allowlist per rule.

use serde::{Deserialize, Serialize};

/// What a rule does on hit.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RuleAction {
    Allow,
    Deny,
}

/// How to interpret the rule's pattern.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PatternKind {
    Exact,
    Prefix,
    Glob,
}

/// The engine's decision on a single URL+method check.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RuleVerdict {
    Allow,
    Deny,
    NoMatch,
}

/// One static URL rule.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UrlRule {
    pub rule_id: String,
    pub pattern: String,
    #[serde(default = "default_kind")]
    pub kind: PatternKind,
    #[serde(default = "default_action")]
    pub action: RuleAction,
    #[serde(default)]
    pub methods: Option<Vec<String>>,
    #[serde(default)]
    pub description: String,
}

fn default_kind() -> PatternKind {
    PatternKind::Exact
}
fn default_action() -> RuleAction {
    RuleAction::Allow
}

impl UrlRule {
    pub fn new(rule_id: impl Into<String>, pattern: impl Into<String>) -> Self {
        Self {
            rule_id: rule_id.into(),
            pattern: pattern.into(),
            kind: PatternKind::Exact,
            action: RuleAction::Allow,
            methods: None,
            description: String::new(),
        }
    }

    pub fn kind(mut self, kind: PatternKind) -> Self {
        self.kind = kind;
        self
    }

    pub fn action(mut self, action: RuleAction) -> Self {
        self.action = action;
        self
    }

    pub fn methods(mut self, methods: Vec<String>) -> Self {
        self.methods = Some(methods.into_iter().map(|m| m.to_uppercase()).collect());
        self
    }

    pub fn description(mut self, d: impl Into<String>) -> Self {
        self.description = d.into();
        self
    }

    fn matches_method(&self, method: &str) -> bool {
        match &self.methods {
            None => true,
            Some(allow) => allow.iter().any(|m| m == &method.to_uppercase()),
        }
    }

    fn matches_url(&self, url: &str) -> bool {
        match self.kind {
            PatternKind::Exact => url == self.pattern,
            PatternKind::Prefix => url.starts_with(&self.pattern),
            PatternKind::Glob => {
                // Use the `glob` crate's Pattern matcher. fnmatch-style
                // `*` and `?` are the most common operator-friendly
                // shapes; the crate also supports `[...]` character
                // classes which is harmless to allow.
                match glob::Pattern::new(&self.pattern) {
                    Ok(p) => p.matches(url),
                    Err(_) => false,
                }
            }
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UrlDecision {
    pub verdict: RuleVerdict,
    pub rule_id: String,
    pub description: String,
    pub method: String,
    pub url: String,
}

impl UrlDecision {
    pub fn no_match(url: &str, method: &str) -> Self {
        Self {
            verdict: RuleVerdict::NoMatch,
            rule_id: String::new(),
            description: String::new(),
            method: method.to_uppercase(),
            url: url.to_string(),
        }
    }

    pub fn allowed(&self) -> bool {
        matches!(self.verdict, RuleVerdict::Allow)
    }
}

/// URL rules engine. Build with [`UrlRulesEngine::new`] and call
/// [`UrlRulesEngine::evaluate`] per request.
#[derive(Clone, Debug)]
pub struct UrlRulesEngine {
    exact: Vec<UrlRule>,
    prefix: Vec<UrlRule>,
    glob: Vec<UrlRule>,
}

impl Default for UrlRulesEngine {
    fn default() -> Self {
        Self::new(Vec::new())
    }
}

impl UrlRulesEngine {
    pub fn new<I: IntoIterator<Item = UrlRule>>(rules: I) -> Self {
        let mut engine = Self {
            exact: Vec::new(),
            prefix: Vec::new(),
            glob: Vec::new(),
        };
        for rule in rules {
            engine.add(rule);
        }
        engine
    }

    pub fn add(&mut self, rule: UrlRule) {
        match rule.kind {
            PatternKind::Exact => self.exact.push(rule),
            PatternKind::Prefix => self.prefix.push(rule),
            PatternKind::Glob => self.glob.push(rule),
        }
    }

    pub fn rule_count(&self) -> usize {
        self.exact.len() + self.prefix.len() + self.glob.len()
    }

    /// Walk the tiers in order (exact > prefix > glob); within each
    /// tier, deny wins over allow. The first decisive verdict wins.
    pub fn evaluate(&self, url: &str, method: &str) -> UrlDecision {
        for tier in [&self.exact, &self.prefix, &self.glob] {
            if let Some(d) = self.evaluate_tier(tier, url, method) {
                return d;
            }
        }
        UrlDecision::no_match(url, method)
    }

    fn evaluate_tier(&self, tier: &[UrlRule], url: &str, method: &str) -> Option<UrlDecision> {
        let mut allow_hit: Option<&UrlRule> = None;
        for rule in tier {
            if !rule.matches_url(url) {
                continue;
            }
            if !rule.matches_method(method) {
                continue;
            }
            if matches!(rule.action, RuleAction::Deny) {
                return Some(UrlDecision {
                    verdict: RuleVerdict::Deny,
                    rule_id: rule.rule_id.clone(),
                    description: rule.description.clone(),
                    method: method.to_uppercase(),
                    url: url.to_string(),
                });
            }
            if allow_hit.is_none() {
                allow_hit = Some(rule);
            }
        }
        allow_hit.map(|r| UrlDecision {
            verdict: RuleVerdict::Allow,
            rule_id: r.rule_id.clone(),
            description: r.description.clone(),
            method: method.to_uppercase(),
            url: url.to_string(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn exact(id: &str, pat: &str, action: RuleAction) -> UrlRule {
        UrlRule::new(id, pat).kind(PatternKind::Exact).action(action)
    }
    fn prefix(id: &str, pat: &str, action: RuleAction) -> UrlRule {
        UrlRule::new(id, pat).kind(PatternKind::Prefix).action(action)
    }
    fn glob_rule(id: &str, pat: &str, action: RuleAction) -> UrlRule {
        UrlRule::new(id, pat).kind(PatternKind::Glob).action(action)
    }

    #[test]
    fn empty_engine_is_no_match() {
        let e = UrlRulesEngine::default();
        assert_eq!(
            e.evaluate("https://example.com/", "GET").verdict,
            RuleVerdict::NoMatch
        );
    }

    #[test]
    fn exact_allow_hit() {
        let e = UrlRulesEngine::new(vec![exact(
            "ok.api",
            "https://api.example.com/v1/health",
            RuleAction::Allow,
        )]);
        let d = e.evaluate("https://api.example.com/v1/health", "GET");
        assert_eq!(d.verdict, RuleVerdict::Allow);
        assert_eq!(d.rule_id, "ok.api");
    }

    #[test]
    fn exact_no_partial_match() {
        let e = UrlRulesEngine::new(vec![exact(
            "only.this",
            "https://api.example.com/v1/health",
            RuleAction::Allow,
        )]);
        assert_eq!(
            e.evaluate("https://api.example.com/v1/health/extra", "GET").verdict,
            RuleVerdict::NoMatch,
        );
    }

    #[test]
    fn exact_deny_wins_in_tier() {
        let e = UrlRulesEngine::new(vec![
            exact("good", "https://api.example.com/v1", RuleAction::Allow),
            exact("bad", "https://api.example.com/v1", RuleAction::Deny),
        ]);
        let d = e.evaluate("https://api.example.com/v1", "GET");
        assert_eq!(d.verdict, RuleVerdict::Deny);
        assert_eq!(d.rule_id, "bad");
    }

    #[test]
    fn prefix_match() {
        let e = UrlRulesEngine::new(vec![prefix(
            "github",
            "https://api.github.com/",
            RuleAction::Allow,
        )]);
        let d = e.evaluate("https://api.github.com/repos/foo/bar", "GET");
        assert_eq!(d.verdict, RuleVerdict::Allow);
        assert_eq!(d.rule_id, "github");
    }

    #[test]
    fn prefix_deny_wins_in_tier() {
        let e = UrlRulesEngine::new(vec![
            prefix("github.all", "https://api.github.com/", RuleAction::Allow),
            prefix(
                "github.admin",
                "https://api.github.com/admin/",
                RuleAction::Deny,
            ),
        ]);
        let d = e.evaluate("https://api.github.com/admin/users", "GET");
        assert_eq!(d.verdict, RuleVerdict::Deny);
        assert_eq!(d.rule_id, "github.admin");
    }

    #[test]
    fn glob_wildcard_matches() {
        let e = UrlRulesEngine::new(vec![glob_rule(
            "any.cdn",
            "https://*.cdn.example.com/*",
            RuleAction::Allow,
        )]);
        assert_eq!(
            e.evaluate("https://images.cdn.example.com/banner.png", "GET").verdict,
            RuleVerdict::Allow,
        );
        assert_eq!(
            e.evaluate("https://example.com/banner.png", "GET").verdict,
            RuleVerdict::NoMatch,
        );
    }

    #[test]
    fn exact_beats_prefix_when_exact_matches() {
        let e = UrlRulesEngine::new(vec![
            exact(
                "specific",
                "https://api.example.com/v1/health",
                RuleAction::Allow,
            ),
            prefix("broad.deny", "https://api.example.com/", RuleAction::Deny),
        ]);
        let d = e.evaluate("https://api.example.com/v1/health", "GET");
        assert_eq!(d.verdict, RuleVerdict::Allow);
        assert_eq!(d.rule_id, "specific");
    }

    #[test]
    fn prefix_falls_through_to_glob() {
        let e = UrlRulesEngine::new(vec![
            prefix("other", "https://other.com/", RuleAction::Allow),
            glob_rule(
                "cdn.match",
                "https://*.cdn.example.com/*",
                RuleAction::Allow,
            ),
        ]);
        let d = e.evaluate("https://images.cdn.example.com/x", "GET");
        assert_eq!(d.verdict, RuleVerdict::Allow);
        assert_eq!(d.rule_id, "cdn.match");
    }

    #[test]
    fn method_filter_only_allows_listed() {
        let e = UrlRulesEngine::new(vec![prefix("github.read", "https://api.github.com/", RuleAction::Allow)
            .methods(vec!["GET".into()])]);
        assert_eq!(
            e.evaluate("https://api.github.com/x", "GET").verdict,
            RuleVerdict::Allow
        );
        assert_eq!(
            e.evaluate("https://api.github.com/x", "POST").verdict,
            RuleVerdict::NoMatch
        );
    }

    #[test]
    fn method_case_insensitive() {
        let e = UrlRulesEngine::new(vec![prefix("ok", "https://x/", RuleAction::Allow)
            .methods(vec!["GET".into()])]);
        assert_eq!(
            e.evaluate("https://x/y", "get").verdict,
            RuleVerdict::Allow
        );
    }

    #[test]
    fn decision_carries_metadata() {
        let e = UrlRulesEngine::new(vec![prefix(
            "github.deny",
            "https://api.github.com/admin/",
            RuleAction::Deny,
        )
        .description("block GitHub admin endpoints")]);
        let d = e.evaluate("https://api.github.com/admin/users", "DELETE");
        assert_eq!(d.verdict, RuleVerdict::Deny);
        assert_eq!(d.rule_id, "github.deny");
        assert_eq!(d.description, "block GitHub admin endpoints");
        assert_eq!(d.method, "DELETE");
        assert_eq!(d.url, "https://api.github.com/admin/users");
        assert!(!d.allowed());
    }

    #[test]
    fn rule_count_reflects_inserted_rules() {
        let e = UrlRulesEngine::new(vec![
            exact("a", "https://x/a", RuleAction::Allow),
            prefix("b", "https://y/", RuleAction::Allow),
            glob_rule("c", "https://*/c", RuleAction::Allow),
            exact("d", "https://x/d", RuleAction::Allow),
        ]);
        assert_eq!(e.rule_count(), 4);
    }

    #[test]
    fn add_after_construction() {
        let mut e = UrlRulesEngine::new(vec![prefix("a", "https://a/", RuleAction::Allow)]);
        e.add(exact("b", "https://b/exact", RuleAction::Allow));
        assert_eq!(e.rule_count(), 2);
        assert_eq!(
            e.evaluate("https://b/exact", "GET").rule_id,
            "b"
        );
    }
}
